uievents/mouse/mousemove_prevent_default_action.tentative.html WPT is failing

aprotyas opened the pull request:
#### 4ed18444c723b19ca7ad1608aae93ca245a38296
<pre>
uievents/mouse/mousemove_prevent_default_action.tentative.html WPT is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=262878">https://bugs.webkit.org/show_bug.cgi?id=262878</a>
rdar://116668701

Reviewed by Aditya Keerthi.

We failed uievents/mouse/mousemove_prevent_default_action.tentative.html
WPT because we were preventing selectionchange or dragstart from
happening on a cancelled mousemove event.

However, mousemove has no default action according to the UI Events spec:
<a href="http://www.w3.org/TR/DOM-Level-3-Events/#event-type-mousemove.">http://www.w3.org/TR/DOM-Level-3-Events/#event-type-mousemove.</a> This
means that preventing the default action on mousemove should not prevent
selection/drag from happening.

To reflect this behavior change, we update the
CapturesDragging::InabilityReason enumeration with a new case --
MouseMoveIsCancelled. We consult m_capturesDragging for this failure
mode in EventHandler::handleMouseMoveEvent before passing a mousemove
event to EventHandler::handleMouseDraggedEvent, which is then
responsible for selection changes and drag operations.

* LayoutTests/fast/text/canceled-mousemove-does-not-prevent-selectionchange-expected.txt: Added.
* LayoutTests/fast/text/canceled-mousemove-does-not-prevent-selectionchange.html: Added.
* LayoutTests/fast/text/selection-is-prevent-defaulted-expected.txt: Removed.
* LayoutTests/fast/text/selection-is-prevent-defaulted.html: Removed.

Rename the layout test and make it more amenable to the fact that
selections happen in both cases now, then update expectations to reflect
our behavior change.

* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mousemove_prevent_default_action.tentative-expected.txt:

Update our test expectations since we now pass the WPT partially.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMouseMoveEvent):
(WebCore::EventHandler::dispatchMouseEvent):
* Source/WebCore/page/EventHandler.h:

Canonical link: <a href="https://commits.webkit.org/269370@main">https://commits.webkit.org/269370@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c46f7ae5c08f39b629b5440f50f810d2621b16f3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22261 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22460 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23333 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24158 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20599 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22508 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26722 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22787 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21608 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22494 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22201 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19315 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25012 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19250 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20171 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26415 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20268 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20401 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24279 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20833 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17723 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20190 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/19980 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5329 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24400 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20780 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/platform/gtk/TestExpectations:
```diff
@@ -2657,6 +2657,8 @@ webkit.org/b/261024 webgl/2.0.y/conformance/renderbuffers/framebuffer-state-rest
 webkit.org/b/261024 webgl/2.0.y/conformance/rendering/color-mask-preserved-during-implicit-clears.html [ Pass Timeout ]
 webkit.org/b/261024 webrtc/utf8-sdp.html [ Pass Timeout ]
 
+fast/text/canceled-mousemove-does-not-prevent-selectionchange.html [ Failure ]
```
What's up with the new failure here?

aprotyas replied:
I'm not 100% sure what happened here. It looks like _some_ type of selection happened with the defaultPrevented mousemove event, but not like on the other ports? I could file a bugzilla for further investigation.

The failure was from this EWS run by the way: https://github.com/WebKit/WebKit/pull/18890#issuecomment-1759693346

```diff
--- /home/ews/worker/GTK-WK2-Tests-EWS/build/layout-test-results/fast/text/canceled-mousemove-does-not-prevent-selectionchange-expected.txt
+++ /home/ews/worker/GTK-WK2-Tests-EWS/build/layout-test-results/fast/text/canceled-mousemove-does-not-prevent-selectionchange-actual.txt
@@ -4,5 +4,3 @@
 Pass if non-selection is visible.
 Try to select this text.
 Pass if non-selection is visible.
-Try to select this text.
-Pass if non-selection is visible.
```

aprotyas replied:
Rewrote the test and removed this failure expectation. Waiting on EWS now! 🤞

aprotyas commented on LayoutTests/platform/mac-wk1/TestExpectations:
```diff
@@ -1086,6 +1086,8 @@ imported/w3c/web-platform-tests/editing/run/multitest.html?5001-6000 [ Slow ]
 imported/w3c/web-platform-tests/editing/run/multitest.html?6001-7000 [ Slow ]
 imported/w3c/web-platform-tests/editing/run/multitest.html?7001-8000 [ Slow ]
 
+imported/w3c/web-platform-tests/uievents/mouse/mousemove_prevent_default_action.tentative.html [ Failure ]
```
We get the desired `selectionchange`/`dragstart` events here, but some of the event ordering is messed up. That is a test issue (and a WPT change proposal has been filed for this)

aprotyas commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -3268,7 +3268,7 @@ webkit.org/b/214299 imported/w3c/web-platform-tests/css/css-ui/outline-with-padd
 webkit.org/b/214498 imported/w3c/web-platform-tests/cors/remote-origin.htm [ Pass ]
 
-fast/text/selection-is-prevent-defaulted.html [ Failure ]
+fast/text/canceled-mousemove-does-not-prevent-selectionchange.html [ Failure ]
```
Pre-existing failure.

aprotyas commented on LayoutTests/platform/wpe/TestExpectations:
```diff
@@ -527,7 +527,8 @@ fast/events/resources/drag-in-frames-left.html [ Skip ]
 fast/events/shift-drag-selection-on-image-triggers-drag-n-drop.html [ Skip ]
 fast/events/shift-drag-selection-on-link-triggers-drag-n-drop.html [ Skip ]
-fast/text/selection-is-prevent-defaulted.html [ Skip ]
+fast/text/canceled-mousemove-does-not-prevent-selectionchange.html [ Failure ]
+imported/w3c/web-platform-tests/uievents/mouse/mousemove_prevent_default_action.tentative.html [ Failure ]
```
No drag support on WPE

aprotyas committed:
uievents/mouse/mousemove_prevent_default_action.tentative.html WPT is failing
https://bugs.webkit.org/show_bug.cgi?id=262878
rdar://116668701

Reviewed by Aditya Keerthi.

We failed uievents/mouse/mousemove_prevent_default_action.tentative.html
WPT because we were preventing selectionchange or dragstart from
happening on a cancelled mousemove event.

However, mousemove has no default action according to the UI Events spec:
http://www.w3.org/TR/DOM-Level-3-Events/#event-type-mousemove. This
means that preventing the default action on mousemove should not prevent
selection/drag from happening.

To reflect this behavior change, we update the
CapturesDragging::InabilityReason enumeration with a new case --
MouseMoveIsCancelled. We consult m_capturesDragging for this failure
mode in EventHandler::handleMouseMoveEvent before passing a mousemove
event to EventHandler::handleMouseDraggedEvent, which is then
responsible for selection changes and drag operations.

* LayoutTests/fast/text/canceled-mousemove-does-not-prevent-selectionchange-expected.txt: Added.
* LayoutTests/fast/text/canceled-mousemove-does-not-prevent-selectionchange.html: Added.
* LayoutTests/fast/text/selection-is-prevent-defaulted-expected.txt: Removed.
* LayoutTests/fast/text/selection-is-prevent-defaulted.html: Removed.

Rename the layout test and make it more amenable to the fact that
selections happen in both cases now, then update expectations to reflect
our behavior change.

* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mousemove_prevent_default_action.tentative-expected.txt:

Update our test expectations since we now pass the WPT partially.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMouseMoveEvent):
(WebCore::EventHandler::dispatchMouseEvent):
* Source/WebCore/page/EventHandler.h:

Canonical link: https://commits.webkit.org/269370@main
